module/state_synchronization/requester/status/status.go

```diff
@@ -0,0 +1,243 @@
+
+	// don't accept entries which have already been notified
+	if entry.Height < s.nextHeight() {
+		return
+	}
+
+	// Enforce a maximum possible cache size. The cache will be effective under normal operation.
+	// However, the cache will be under utilized if there is a large gap of unfetched heights.
+	if entry.Height >= (s.lastNotified + s.maxCachedEntries) {
+		entry.ExecutionData = nil
+	}
+
+	heap.Push(s.heap, entry)
+
+	// Only track the highest height
+	if entry.Height > s.lastReceived {
+		s.lastReceived = entry.Height
+	}
+
+	// Pause the download workers if the lastNotified is more than maxSearchAhead blocks
+	// behind. Since workers will not return until they have successfully downloaded the
+	// ExecutionData, this will effectively pause new downloads, and allow existing ones
+	// to complete (including the next to notify).
+	if s.shouldPause() {
+		s.ingester.Pause()
+	}
+}
+
+// Halt marks the requester as halted. This is persisted between restarts
+func (s *Status) Halt() {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	s.halted = true
+
+	if err := s.progress.SetHalted(true); err != nil {
+		s.log.Error().Err(err).Msg("failed to persist halted state")
+	}
+}
+
+// LastNotified returns the last block height that was returned from nextNotification
+func (s *Status) LastNotified() uint64 {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	return s.lastNotified
+}
+
+// Halted returns true if the requester has been halted
+func (s *Status) Halted() bool {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	return s.halted
+}
+
+// NextNotificationHeight returns the next block height waiting to be notified.
+// This height may not be available yet.
+func (s *Status) NextNotificationHeight() uint64 {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	return s.nextHeight()
+}
+
+// nextNotification returns the BlockEntry for the next notification to send and marks it as notified.
+// If the next notification is available, it returns the BlockEntry and true. Otherwise, it returns
+// nil and false.
+func (s *Status) nextNotification() (*BlockEntry, bool) {
+	var entry *BlockEntry
+	next := s.nextHeight()
+
+	// Remove any entries for the next height or below. This ensures duplicate or unexpected heights
+	// below the tracked heights are removed. If there are duplicate entries for the same height, the
+	// last entry popped from the heap will be returned.
+	for s.heap.Len() > 0 && s.heap.PeekMin().Height <= next {
+		entry = heap.Pop(s.heap).(*BlockEntry)
+	}
+
+	if entry == nil || entry.Height != next {
+		return nil, false
+	}
+
+	// Now we have the next height to notify
+	s.lastNotified = entry.Height
+
+	// Resume processing if notifications have caught up
+	// Resume is a noop if the block consumer is already running
+	if !s.shouldPause() {
+		s.ingester.Resume()
+	}
+
+	return entry, true
+}
+
+func (s *Status) nextHeight() uint64 {
+	return s.lastNotified + 1
+}
+
+func (s *Status) shouldPause() bool {
+	return s.lastReceived >= s.lastNotified+s.maxSearchAhead
+}
```

Comment:
This is a feature that the jobqueue didn't implement, but I think is not hard to add. 

We can simply add a config and update [this for loop condition](https://github.com/onflow/flow-go/blob/petera/execution-data-requester/module/jobqueue/consumer.go#L301) to be:
```
for i := processedIndex + 1; processing < maxProcessing && len(processings) <= maxSearchAhead; i++ {
```

The `processings` map contains both processed jobs that are in higher heights, as well as the jobs not completed. Since the heights in processings are consecutive, checking the count of that is enough.

Furthermore, we can simplify by getting rid of the [Pause and Resume](https://github.com/onflow/flow-go/blob/petera/execution-data-requester/module/jobqueue/consumer.go#L134-L152) methods, because the jobqueue is able to pause by it own. No need to let external caller determine when to pause.

Reply:
updated in 27ec92c